src/table.rs

```diff
@@ -2552,15 +2580,112 @@ where
     T: fmt::Debug,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_list()
-            .entries(Iter {
-                inner: self.inner.clone(),
-                marker: PhantomData,
-            })
-            .finish()
+        f.debug_list().entries(self.iter()).finish()
+    }
+}
+
+/// An unsafe version of [`IterMut`] which is *variant* over `T`.
+///
+/// This is used for implementations of mutable iterators where mutation is only
+/// allowed on part of the value in the table. For example, a mutable iterator
+/// for a map may return an immutable key alongside a mutable value, even though
+/// these are both stored inside the table.
+///
+/// # Safety
+///
+/// In order to correctly use this iterator, it should be wrapped in a safe
+/// iterator struct with the appropriate [`PhantomData`] marker to indicate the
+/// correct variance.
+///
+/// For example, a [`hash_map::IterMut`] implementation correctly returning
+/// a variant key, and an invariant value:
+///
+/// ```rust
+/// pub struct IterMut<'a, K, V> {
+///     inner: hash_table::IterMut<'a, (K, V)>,
+///     // Variant over keys, invariant over values
+///     marker: PhantomData<(&'a K, &'a mut V)>,
+/// }
+/// impl<'a, K, V> Iterator for IterMut<'a, K, V> {
+///     // Immutable keys, mutable values
+///     type Item = (&'a K, &'a mut V);
+///
+///     fn next(&mut self) -> Option<Self::Item> {
+///         // Even though the key is mutably borrowed here, this is sound
+///         // because we never actually mutate the key before yielding the
+///         // immutable reference
+///         let (ref key, ref mut val) = self.inner.next()?;
+///         Some((key, val))
```

Comment:
Since this is just an example, and not part of the actual implementation, I chose to use a clearer implementation instead of the one we actually use.